app/lib/features/cross_signing/providers/notifiers/verification_notifiers.dart

```diff
@@ -0,0 +1,239 @@
+import 'dart:async';
+
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+import 'package:logging/logging.dart';
+import 'package:riverpod/riverpod.dart';
+
+final _log = Logger('a3::cross_signing::notifier');
+
+class VerificationState {
+  final String stage;
+  final VerificationEvent? event; // null for verification.init
+  final bool isVerifier; // whether this device requested verification
+  final String? flowId; // if not null, verification flying now
+  final bool keysExchanged;
+
+  const VerificationState({
+    required this.stage,
+    this.event,
+    required this.isVerifier,
+    this.flowId,
+    required this.keysExchanged,
+  });
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! VerificationState) {
+      _log.info('is not VerificationState');
+      return false;
+    }
+    if (other.stage != stage) {
+      _log.info('stage is not same');
+      return false;
+    }
+    // don't compare isVerifier & flowId & keysExchanged
+    // because they are changed via changeState in ref.listen
+    return true;
+  }
+
+  @override
+  int get hashCode => stage.hashCode;
+}
+
+class VerificationNotifier extends StateNotifier<VerificationState> {
+  final Ref ref;
+  final Client client;
+
+  late Stream<VerificationEvent>? _listener;
+  late StreamSubscription<VerificationEvent>? _poller;
+
+  VerificationNotifier({
+    required this.ref,
+    required this.client,
+  }) : super(
+          const VerificationState(
+            stage: 'verification.init',
+            isVerifier: false,
+            keysExchanged: false,
+          ),
+        ) {
+    _init();
+  }
+
+  void _init() {
+    _listener = client.verificationEventRx(); // keep it resident in memory
+    _poller = _listener?.listen(_handleEvent);
+    ref.onDispose(() => _poller?.cancel());
+  }
+
+  void _handleEvent(VerificationEvent event) {
+    String eventType = event.eventType();
+    _log.info('$eventType - flow_id: ${event.flowId()}');
+    switch (eventType) {
+      /* request_verification_handler: verifier makes request */
+      case 'VerificationRequestState::Created':
+        state = VerificationState(
+          stage: 'request.created',
+          event: event,
+          isVerifier: state.isVerifier,
+          flowId: state.flowId,
+          keysExchanged: state.keysExchanged,
+        );
+        break;
```

Comment:
pretty sure @gtalha07 is happy to teach you about `freezed` as well at some point, as it would make this code a lot less verbose... but IMHO this is fine for now ;)